department-of-veterans-affairs/vets-website
# 19516 Email Signup Widget

## Are you removing, renaming or moving a folder in this PR?

### :warning: TeamSites :warning:
Did you change site-wide styles, platform utilities or other infrastructure?

## Summary

Created a React widget for the email signup form on the VA.gov homepage. This replaces the form that was in content-build.

Figma design: https://www.figma.com/design/60PPoOSpKiXwmNHT01C09J/Home-Page-Updates?node-id=38-3849&node-type=canvas&t=Q63YjY6qharfzzqi-0

## Related issue(s)
- https://github.com/department-of-veterans-affairs/va.gov-cms/issues/19516

## Testing done
content-build has a feature toggle that controls whether this new React widget displays or the original content-build email signup form.

When the feature toggle is **ON**, the original content-build form should display.

In staging and production, the feature toggle is **ON**. In [this Tugboat](https://cms-g1rglcuobmlvgewnhny9uo7f6vigntwi.demo.cms.va.gov/), the feature toggle is **OFF**.

To test both scenarios, first:
1. Point your `env` DRUPAL_ADDRESS in content-build to `https://staging.cms.va.gov` on the `main` branch
2. Run content-build
3. Run vets-website on this branch (`19516-email-signup-widget`)
4. Verify that you see the original email signup form. You can tell easily because it won't have `Required` or a character counter on it
5. Verify that the Veteran banner still appears. This banner is part of the email signup template in content-build; this is a regression check

<img width="400" alt="Screenshot 2024-11-07 at 7 16 36 AM" src="https://github.com/user-attachments/assets/be5a0f68-0c16-4420-9847-0cdd16f99e2e">

Then:
1. Point your `env` DRUPAL_ADDRESS in content-build to `https://cms-g1rglcuobmlvgewnhny9uo7f6vigntwi.demo.cms.va.gov` on the `main` branch
2. Run content-build
3. Run vets-website on this branch (`19516-email-signup-widget`)
4. Verify that you see the new React widget form. You can tell easily because it will have `Required` and a character counter on it
5. Verify that the Veteran banner still appears. This banner is part of the email signup template in content-build; this is a regression check
6. Check against screenshots below and [the Figma design](https://www.figma.com/design/60PPoOSpKiXwmNHT01C09J/Home-Page-Updates?node-id=38-3849&node-type=canvas&t=Q63YjY6qharfzzqi-0) across breakpoints

### Testing behavior
1. The email form should not submit when the input is blank
2. The email form should not submit when an invalid email is entered (invalid format / special characters)
3. The email form should submit when the input is filled with a valid email and the "Sign up" button is clicked
4. The email form should submit when the input is filled with a valid email and the Enter key is hit while the input is focused
5. The email input should show a yellow focus ring when the user is typing an email
6. When there is an error with email input, the screenreader should read out the error and not allow form submission
7. All of the above behaviors should be the same across breakpoints

## Screenshots

<details><summary>Desktop</summary>







#### Fresh form
<img width="1000" alt="Screenshot 2024-11-07 at 1 42 15 PM" src="https://github.com/user-attachments/assets/2aead7df-29a7-4f84-9218-1a20b7832348">

#### Form with email address filled out
<img width="1000" alt="Screenshot 2024-11-07 at 1 42 23 PM" src="https://github.com/user-attachments/assets/eae46109-9140-49bd-9e1f-a46ad48dd2a2">

#### Form in error state
<img width="1001" alt="Screenshot 2024-11-07 at 1 42 30 PM" src="https://github.com/user-attachments/assets/093b1a35-0209-40b5-913e-bd484fd7e8bb">

#### GovDelivery form; VA.gov form has been submitted by pressing enter

<img width="1000" alt="Screenshot 2024-11-07 at 7 09 00 AM" src="https://github.com/user-attachments/assets/06a76548-076d-4c2c-85d3-bfb60cb5e6cb">

#### GovDelivery form; VA.gov form has been submitted by clicking "Sign up"

<img width="1000" alt="Screenshot 2024-11-07 at 7 12 16 AM" src="https://github.com/user-attachments/assets/e035cce2-ccca-4f59-82d3-9c48ddf4e440">

</details>

<details><summary>Tablet</summary>

#### Fresh form
<img width="768" alt="Screenshot 2024-11-07 at 1 41 49 PM" src="https://github.com/user-attachments/assets/9740d22b-37fc-42c9-b814-5d7287702d16">

#### Form with email address filled out
<img width="768" alt="Screenshot 2024-11-07 at 1 41 57 PM" src="https://github.com/user-attachments/assets/398d5dc3-f1d6-47a4-93d3-5a01c8f759eb">

#### Form in error state
<img width="769" alt="Screenshot 2024-11-07 at 1 42 05 PM" src="https://github.com/user-attachments/assets/fef67a78-4c8d-4480-8bf1-862e6127be22">

</details>

<details><summary>Mobile</summary>

#### Fresh form
<img width="400" alt="Screenshot 2024-11-07 at 1 41 16 PM" src="https://github.com/user-attachments/assets/126e4192-025a-49c7-ac63-252f619da12a">

#### Form with email address filled out
<img width="400" alt="Screenshot 2024-11-07 at 1 41 25 PM" src="https://github.com/user-attachments/assets/59597fc2-c157-45c6-adaf-ace9a3df3c81">

#### Form in error state
<img width="400" alt="Screenshot 2024-11-07 at 1 41 35 PM" src="https://github.com/user-attachments/assets/412c7fce-8e5d-403b-a59b-aaa6953122d9">


</details>

## What areas of the site does it impact?
Homepage - email signup only

## Acceptance criteria
- [x] Create a React widget in vets-website for the form to include:
  - [x] Input validation built in from the web component
  - [x] Input validation messaging aligning with the Figma design
- [x] Add unit tests for the new React widget as appropriate/capable
